core/src/main/scala/org/apache/spark/rdd/HadoopRDD.scala

```diff
@@ -157,20 +157,23 @@ class HadoopRDD[K, V](
       if (conf.isInstanceOf[JobConf]) {
         logDebug("Re-using user-broadcasted JobConf")
         conf.asInstanceOf[JobConf]
-      } else if (HadoopRDD.containsCachedMetadata(jobConfCacheKey)) {
-        logDebug("Re-using cached JobConf")
-        HadoopRDD.getCachedMetadata(jobConfCacheKey).asInstanceOf[JobConf]
       } else {
-        // Create a JobConf that will be cached and used across this RDD's getJobConf() calls in the
-        // local process. The local cache is accessed through HadoopRDD.putCachedMetadata().
-        // The caching helps minimize GC, since a JobConf can contain ~10KB of temporary objects.
-        // Synchronize to prevent ConcurrentModificationException (SPARK-1097, HADOOP-10456).
-        HadoopRDD.CONFIGURATION_INSTANTIATION_LOCK.synchronized {
-          logDebug("Creating new JobConf and caching it for later re-use")
-          val newJobConf = new JobConf(conf)
-          initLocalJobConfFuncOpt.foreach(f => f(newJobConf))
-          HadoopRDD.putCachedMetadata(jobConfCacheKey, newJobConf)
-          newJobConf
+        val newJobConf = HadoopRDD.getCachedMetadata(jobConfCacheKey)
```

Comment:
You are using two val with the same name here, normally we don't recommend that this way. How about follow vanzin's suggestion in the JIRA issue, like this:
```
Option(HadoopRDD.getCachedMetadata(jobConfCacheKey)).getOrElse(
     HadoopRDD.CONFIGURATION_INSTANTIATION_LOCK.synchronized {
          ......
})
```

Reply:
I could do that, but then I would have to remove the `logDebug("Re-using cached JobConf")` statement. Unless there is some way to get around that in Scala.

Otherwise, I can remove the val name to `cachedJobConf`.

Reply:
You could do that without the val:

```
Option(HadoopRDD.getCachedMetadata(jobConfCacheKey))
  .map { conf =>
     logDebug(...)
     conf.asInstanceOf[...]
  }
  .getOrElse {
     // code in the "else" block
  }
```

Reply:
Thanks for the tip. Updated the patch.